### PR TITLE
Rubocop fixes for #653

### PR DESCRIPTION
**Pull request by Justin-W**
This PR resolves many of the rubocop cop offenses (#653) which were previously-ignored due to the auto-generated config in .rubocop_todo.yml.

Almost all of the changes are "pure code refactoring", with zero impact to the behavior of the modified code (the notable exception being 1 unit test, which was found to be previously-broken, and is now fixed).

All of the fixed/excluded offenses were resolved incrementally (one cop at a time, in the order indicated below), by either auto-fixing the manually re-enabled offenses or else explicitly disabling/excluding (inline or in the .rubocop.yml, instead of in .rubocop_todo.yml) the previously-ignored offenses.

After each incremental change (whether auto-fixed by rubocop or manually fixed/resolved), the incremental code changes were re-reviewed (for both correctness and readability), and the unit test suite was re-executed (to verify no regressions were introduced). Where either the code review and/or unit tests indicated that the auto-fixed changes were unacceptable, the auto-fixed changes were discarded and manual fixes/resolution were applied instead.

Auto-fixes were performed via the `rubocop -x`, `rubocop -a`, or `rubocop -A` commands (in that order), depending on the auto-correct requirements for each individual cop.

# Checklist:

Please check each of the boxes below for which you have completed the corresponding task:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All unit tests pass locally (after my changes)
- [x] Rubocop reports zero offenses (after my changes)

Please include below the summary portions of the output from the following 2 scripts:

  ```
  pushd src/bosh_azure_cpi
    ./bin/test-unit
    ./bin/rubocop_check
  popd
  ```

  _NOTE:_ Please see how to setup dev environment and run unit tests in docs/development.md.

### Unit Test output:

> 1013 examples, 0 failures
Coverage report generated for RSpec to /Users/justinw/go/src/github.com/cloudfoundry/bosh-azure-cpi-release/src/bosh_azure_cpi/coverage. 4234 / 4253 LOC (99.55%) covered.

### Rubocop output:

> 187 files inspected, no offenses detected

# Changelog

* Fixed 1 broken unit test ("Bosh::AzureCloud::Cloud#set_disk_metadata should raise a NotSupported error")
* Fixed many previously-TODO rubocop offenses (#653)
* Explicitly disabled some previously-TODO rubocop offenses (#653)

### CODE REVIEW

**Comment by Justin-W** [comment]
@klakin-pivotal Regarding the number of commits: I wasn't quite sure how much I should squash the commits in this PR.

On the one hand, each rubocop cop's offenses represent a distinct set of issues to be resolved (and the fixes typically involve similar code changes, making them easy to review toether).

On the other hand, all of the changes (whether auto-fixed or manual) have the same objectives (i.e. improve code correctness, maintainability, readability, etc.), but squashing together the fixes for multiple cops' offenses makes the code review more cognitively difficult.

Therefore, I've prepared 3 different branches as options for you:

- A: https://github.com/Justin-W/bosh-azure-cpi-release/tree/rubocop-fixes-a: This branch has all 66 of the original incremental commits, with auto-fixes mostly grouped together (to the extent possible without nasty re-working due to major merge conflicts from rebasing).
- B: https://github.com/Justin-W/bosh-azure-cpi-release/tree/rubocop-fixes-b: This branch squashes the consecutive single-cop auto-fix commits into multi-cop auto-fix commits, reducing the commit count from 66 to 22, based on the assumption that PR reviewers may not care to review "safe" auto-corrections made by rubocop independently, but may want to separately review (and possibly ask for revisions/reverts of) any of the individual manually-applied fixes/changes.
- C: https://github.com/Justin-W/bosh-azure-cpi-release/tree/rubocop-fixes-c: This squashes all 66 original commits down into 1 commit. All or nothing. :)

This PR is currently based on branch 'B', for the reasons noted above (and since it seemed like the probably-best option based on your prior feedback for a different PR). However, if you would prefer more or fewer commits, I'd be happy to revise the PR to the HEAD of either of the other 2 branches.

**Comment by klakin-pivotal** [comment]
So, I'm probably not the fellow you should be asking to review PRs that change `rubocop` stuff. Because it has never told me to change anything that has been actually useful, (and has -at times- demanded that I change _many_ things that have made my programs _substantially_ more difficult to read and understand) I _hate_ it with a _burning passion_.

Anyway. A few things:

1) If I were reviewing this PR, and if I didn't _hate_ `rubocop` I would prefer that all the RuboCop changes be squashed into a single commit, with the reasoning behind any changes that you've made in the commit message. So, if you feel that the the 22-commit PR is a good way to break out the changes, _I_ would prefer that the reasoning described in those 22 commit messages be put into the message of a single squash commit.
1a) I also would not include the text from the `SQUASHED COMMIT HISTORY:` part onwards. In my opinion, the commit messages in this section don't contain enough information to be worth keeping around. Any information that they _would_ contain should be described with some prose in the commit message.
2) I notice that _most_ of the `Offense Count` counts (in terms of number of lines that contain the phrase "Offense count: ") in `src/bosh_azure_cpi/.rubocop_todo.yml` increased. What's up with that? (Though, to be fair, some of the ones that decrased, decreased a _ton_.)
3) I'll restate this, but more explicitly: I probably should not be reviewing this PR... I _really_, _really_, dislike `rubocop`, and wouldn't mind if it was suddenly erased from existence. ;)

So, yeah.

Someone who's not me should actually review this PR, and if you dislike what I've said here, please do ignore it!

(@Justin-W, can you request a review from someone, or do you need me to select someone? That is, do you see a thing in the upper-right-ish of the UI that says `Reviewers ... Suggestions`, and maybe has three folks in the list? The UI I'm talking about seems to be pretty close to what's described in this blog post: <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review>)

**Comment by Justin-W** [comment]
@klakin-pivotal Thanks for the response. Answers below.

> So, I'm probably not the fellow you should be asking to review PRs that change `rubocop` stuff. Because it has never told me to change anything that has been actually useful, (and has -at times- demanded that I change _many_ things that have made my programs _substantially_ more difficult to read and understand) I _hate_ it with a _burning passion_.

Lol. Ok, fair enough. And I've had some similar frustrations, but have found that it does occasionally expose real issues (like the test which was broken but incapable of failing, and which I've fixed in this PR). And such "hidden bugs" were the primary reason for even attempting this PR (with code style consistency across the project being secondary).

> 1. If I were reviewing this PR, and if I didn't _hate_ `rubocop` I would prefer that all the RuboCop changes be squashed into a single commit, 
> ... 
> Any information that they _would_ contain should be described with some prose in the commit message.

I'm fine with that. But it sounds like you think maybe the "different reviewer" should give an opinion first? I really don't care either way, other than trying to minimize additional/wasted time spent on this PR.

> 2. I notice that _most_ of the `Offense Count` counts (in terms of number of lines that contain the phrase "Offense count: ") in `src/bosh_azure_cpi/.rubocop_todo.yml` increased. What's up with that? (Though, to be fair, some of the ones that decrased, decreased a _ton_.)

I checked, and it looks like the ones that increased slightly were mostly increased because the `src/bosh_azure_cpi/.rubocop_todo.yml`  file wasn't updated by #651, and the new code added by that PR increased those offense counts because the new code matched the project's existing (rubocop-cop-offending) code style (and mainly in files which already contained similar offenses before #651). So those increased counts should be viewed as corrections to the pre-PR counts, not as new offenses from this PR.

> ...
> Someone who's not me should actually review this PR, and if you dislike what I've said here, please do ignore it!
> 
> (@Justin-W, can you request a review from someone, or do you need me to select someone? 

I do not have the permission to assign reviewers within this project, so you (or @rkoster) would need to make such assignments. (I have it on other orgs/projects, but not for this one.) 

Also, I only made this whole set of "fixes" with the intent to improve the state of the project. So if there are serious concerns about whether it actually does so, I'm open to hearing that.

**Review by beyhan — COMMENTED** [review]
@Justin-W  thank you for looking into this. The change looks good to me. I just have one comment about one refactoring (change) which I couldn't follow.

File: src/bosh_azure_cpi/lib/cloud/azure/telemetry/telemetry_event.rb [hunk 1]
```diff
@@ -19,8 +19,8 @@ def to_hash
       { 'name' => @name, 'value' => @value }
     end
 
-    def to_json
-      to_hash.to_json
+    def to_json(*args)
```
[the remark this answers is not in the document]
Comment (Justin-W) [reply src/bosh_azure_cpi/lib/cloud/azure/telemetry/telemetry_event.rb:22]: This change was needed to resolve a `Lint/ToJSON` cop offense.
See: https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Lint/ToJSON

`#to_json` methods are supposed to support pass-through of the optional `opts` arg supported by the following JSON module methods:
- [JSON#generate](https://ruby-doc.org/stdlib-2.6.3/libdoc/json/rdoc/JSON.html#method-i-generate)
- [JSON#fast_generate](https://ruby-doc.org/stdlib-2.6.3/libdoc/json/rdoc/JSON.html#method-i-fast_generate)
- [JSON#pretty_generate](https://ruby-doc.org/stdlib-2.6.3/libdoc/json/rdoc/JSON.html#method-i-pretty_generate)

These optional args can be used to configure certain JSON formatting options, such as tabs vs spaces, space chars per indent level, etc. See the methods' docs for details.

The change I made was simply to allow the method to accept a list of args, and pass the whole args list through to the `Hash#to_json` method. Any callers of the modified method which don't pass args will see no difference in behavior. But callers can now optionally pass an `opts` hash to configure the returned JSON's formatting (which resolves the `Lint/ToJSON` cop offense).

File: src/bosh_azure_cpi/lib/cloud/azure/telemetry/telemetry_event.rb [hunk 2]
```diff
@@ -116,8 +116,8 @@ def to_hash
       }
     end
 
-    def to_json
-      to_hash.to_json
+    def to_json(*args)
```
[the remark this answers is not in the document]
Comment (Justin-W) [reply src/bosh_azure_cpi/lib/cloud/azure/telemetry/telemetry_event.rb:119]: Exactly the same reason as [explained above](https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/658#discussion_r798848625).

File: src/bosh_azure_cpi/lib/cloud/azure/telemetry/telemetry_event.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (beyhan) [reply src/bosh_azure_cpi/lib/cloud/azure/telemetry/telemetry_event.rb:22]: Thank you for the details. It makes sense.

**Review by beyhan — APPROVED** [review]

**Review by rkoster — APPROVED** [review]
Wow, that was a lot of changes, to review. But apart from the things Beyhan already asked about, everything made sense.

**Comment by rkoster** [comment]
Thanks! @Justin-W

**Pull request merged**